### ISSUE TITLE
Typographic quotes should not be string termiators

### ISSUE THREAD

**Issue by 89z** [PowerShell/PowerShell#16084]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

If I have a file with a typographical quote:

~~~
DON’T LOOK UP.txt
~~~

### Expected behavior

```console
and I try to tab complete
```

### Actual behavior

```console
~~~
Get-Content '.\DON’T LOOK UP.txt'
~~~

that looks correct enough, but for some reason, PowerShell thinks a typographic quote is also a **programmatic quote**, and so PowerShell thinks the string ends of the typographic quote. This issue can be fixed two ways:

1. Stop treating typographic quotes as special. Only single quote (U+0027) or double quote (U+0022) should be recognized as string terminators.

2. If typographic quotes are going to be used as programmatic quotes, then in the example above, tab complete should use double quotes instead.
```

### Environment data

```powershell
~~~
PSVersion                      7.1.4
PSEdition                      Core
GitCommitId                    7.1.4
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
~~~
```

**Comment by jborean93** [comment]
> If typographic quotes are going to be used as programmatic quotes, then in the example above, tab complete should use double quotes instead.

I feel like this should be the way to go forward, tab completion should be able to handle the various quoting types. The other option has the potential to fail existing code even if it's a pretty rare occurrence.

**Comment by 89z** [comment]
> I feel like this should be the way to go forward, tab completion should be able to handle the various quoting types. The other option has the potential to fail existing code even if it's a pretty rare occurrence.

I dont think I should have even given this as an option, as its so obviously wrong.

If you consider the current 10 most used programming languages, not a single one does this. We are not typsetting a printable document. We are doing programming (even if its shell programming). In that context, giving any special meaning to typographic quotes is just wrong.

Trying to fix that error, by adding even more code around the "typographic quote string terminator" code would just be digging a hole even deeper. Please dont.

**Comment by jborean93** [comment]
> I dont think I should have even given this as an option, as its so obviously wrong.

You may thing it's wrong but others may feel differently. I've seen a few websites in the past where it turns text into the "smart" variant and things like `-` becomes an em-dash and quotes become their "smart" variants. Storing code in a word document, while not very ideal, is something I've seen people do and word typically likes to change the characters to the smart variant as well.

Just wanting to put it out there that if the decision is made to stop treating [these chars](https://github.com/PowerShell/PowerShell/blob/b7cb335f03fe2992d0cbd61699de9d9aafa1d7c1/src/System.Management.Automation/engine/parser/CharTraits.cs#L18-L24) as quotes then we need to be aware that this could break existing code. This has been in place since the Windows PowerShell days so it's something that's lived with PowerShell since the early days.

Of course I'm just a lone voice here, it's up to the working group to decide the next step forward :)

**Comment by 89z** [comment]
In my opinion, it shouldnt be PowerShells job to sanitize input from Microsoft Word. In fact generally speaking I dont think Microsoft Word should be mentioned in the same breath as PowerShell, as they are wholly different platforms. That fact that its technically possible for someone to copy some arbitrary text from Microsoft Word and paste into PowerShell isnt something we should be optimizing for. So not only has pointless complexity been added (from the links you gave), but that complexity has also introduced bugs.

If the PowerShell group insist on supporting Microsoft Word, then fine thats their business. But its just making it more difficult on them, as the code is more complex, and more prone to error. For example without that code you linked, the issue I posted here would never exist.

**Comment by 89z** [comment]
> A lot of blogs did as well, at least at the time. I believe it was to make it easier to copy and paste examples from blogs like that. PowerShell early on had a lot of "wouldn't it be cool if __ just _worked_?" that was not vetted to the same degree it would be today.

Look, all I care about is that the issue is fixed. Personally I think all the justifications Ive seen so far are ridiculous, and the idea that people are not purposefully using typographic quotes **in scripts** is disturbing. Its also upsetting that not only was this (in my opinion) pointless change made, but it was done in such a careless way, that even the most basic testing would have caught.

At any rate, as youve said, the ship has sailed. So if the fix has to be "add double quotes instead" in the original example above, so be it.

**Comment by SeeminglyScience** [comment]
> Its also upsetting that not only was this (in my opinion) pointless change made, but it was done in such a careless way, that even the most basic testing would have caught.

Do you mean there was a specific recent break in the way tab completion works? Or are you referring to acceptance of curved quotes in general which has always been the case (or at least since version 2).

**Comment by 89z** [comment]
> Do you mean there was a specific recent break in the way tab completion works? Or are you referring to acceptance of curved quotes in general which has always been the case (or at least since version 2).

@SeeminglyScience why are you acting as if you havent even read the original issue? I would advise you to do so, before commenting futher. In that future, I would suggest that you read the original issue before commenting. Its at the top of this page, but I will link it for your convenience:

https://github.com/PowerShell/PowerShell/issues/16084#issue-993881924

**Comment by SeeminglyScience** [comment]
Yes I read the issue.

> **pointless change made, but it was done in such a careless way, that even the most basic testing would have caught.**

The above implies that something worked previously and no longer works.  That's what I'm asking when I say this:

> Do you mean there was a specific recent break in the way tab completion works?

It's important to know if it's a regression or just a bug that hasn't been filed until now.

**Comment by 89z** [comment]
@SeeminglyScience Ive blocked you, you wont be able to comment further.

Despite you making 3 comments, youve so far failed to add anything to the conversation. This issue isnt "typographic quotes are bad, prove me wrong", and it never was. Ive updated the title, to hopefully better convey that fact.

The real issue, is that the current code around string quoting is wrong, because its easily broken by tab completion, as I demonstrated in the original issue. Ideally I would like to see this fixed by simply ripping out all the code for typographic quotes, but I can understand why that might not be an option. Regardless, the original issue is a valid on, which if not solved by removing support to typographic quotes, still needs to be solved another way (double quoting, etc).

Ive made my personal opinions on typographic quotes very clear. I dont (and never did) want to hear anyone elses opinion on that point. What I would like, is for the original issue to be resolve (however that might be).

**Comment by alexandair** [comment]
@89z You won't go far with that kind of attitude in this community. Your tone isn't nice from the start. People here try to help and understand your issue.

**Comment by 89z** [comment]
@alexandair thank you for the notice. Please let me know if you have any on topic comments.

**Comment by rjmholt** [comment]
The fix we'd want to make is here:

https://github.com/PowerShell/PowerShell/blob/55ccbb6803162965eead3a08aace37f11ea7630b/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L4442-L4472

Everywhere there where quotes are tested for and replaced, we should be looking for all the kinds of quotes the tokeniser accepts.

In the tab completion code, the most efficient and readable way to do that is probably to use a regex that tests for all the quote types.

We should also be doing this for double quotes.

**Comment by iSazonov** [comment]
> The fix we'd want to make is here:

~15 such patterns are in the file :-)

**Comment by rjmholt** [comment]
> ~15 such patterns are in the file :-)

Yeah the way we generally handle quotation in the completers is badly done and ideally we would fix it all.

But I debugged the actual reported issue and that's the specific part of the code we'd want to change.

**Comment by 89z** [comment]
bump

**Comment by GigaScratch** [comment]
Do we have a record of all characters that can cause this? Writing a script that tests every unicode character for tab completion seems like overkill but I can't think of another way.
